nemo/collections/nlp/modules/common/transformer/transformer_generators.py

```diff
@@ -0,0 +1,324 @@
+        encoder_input_mask=None,
+        decoder_mems_list=None,
+        pos=0,
+    ):
+        """
+        One step of autoregressive output generation.
+
+        Args:
+            decoder_input_ids: starting sequence of tokens to generate from;
+                if None, generation will start from a batch of <bos> tokens
+            encoder_hidden_states: output of the encoder for conditional
+                sequence generation; if None, generator will use unconditional
+                mode (e.g., language modeling)
+            encoder_input_mask: input mask used in the encoder
+            decoder_mems_list: list of size num_layers with cached activations
+                of sequence (x[1], ..., x[k-1]) for fast generation of x[k]
+            pos: starting position in positional encoding
+        """
+
+        decoder_hidden_states = self.embedding.forward(decoder_input_ids, start_pos=pos)
+        decoder_input_mask = mask_padded_tokens(decoder_input_ids, self.pad).float()
+        # TODO: make sure float() work with mixed precision
+
+        if encoder_hidden_states is not None:
+            decoder_mems_list = self.decoder.forward(
+                decoder_hidden_states,
+                decoder_input_mask,
+                encoder_hidden_states,
+                encoder_input_mask,
+                decoder_mems_list,
+                return_mems=True,
+            )
+        else:
+            decoder_mems_list = self.decoder.forward(
+                decoder_hidden_states, decoder_input_mask, decoder_mems_list, return_mems=True
+            )
+        log_probs = self.log_softmax.forward(decoder_mems_list[-1])
+        return log_probs, decoder_mems_list
+
+    def _prepare_for_search(self, decoder_input_ids=None, encoder_hidden_states=None):
+        """
+        Helper function which defines starting sequence to begin generating
+        with and maximum allowed number of tokens to be generated.
+        """
+
+        batch_size = self.batch_size
+
+        # for encoder-decoder generation, maximum length of generated sequence
+        # is min(max_sequence_length, src_len + max_delta_length)
+        if encoder_hidden_states is not None:
+            batch_size, src_len, _ = encoder_hidden_states.size()
+            max_seq_length = min(self.max_seq_length, src_len + self.max_delta_len)
+        else:
+            max_seq_length = self.max_seq_length
+
+        # if no input is provided, start with the batch of <bos> tokens
+        if decoder_input_ids is not None:
+            tgt = decoder_input_ids
+            batch_size, tgt_len = decoder_input_ids.size()
+        else:
+            tgt = torch.zeros(batch_size, 1).long().fill_(self.bos).to(self.device)
+            tgt_len = 1
+        max_generation_length = max_seq_length - tgt_len
+
+        return tgt, batch_size, max_generation_length
+
+    def forward(self, decoder_input_ids=None, encoder_hidden_states=None, encoder_input_mask=None):
```

Comment:
could you add docstring

Reply:
Do we need to add docstrings to all forward methods? I did not see it in other models.

Reply:
we are adding them to user-exposed APIs